utils.go

```diff
@@ -14,14 +31,185 @@ type Storage struct {
 
 // String implements Storager.String
 func (s *Storage) String() string {
-	panic("implement me")
+	return fmt.Sprintf("Storager ftp {URL: %s, User: %s, WorkDir: %s}", s.url, s.user, s.workDir)
 }
 
 // NewStorager will create Storager only.
 func NewStorager(pairs ...types.Pair) (types.Storager, error) {
-	panic("implement me")
+	return newStorager(pairs...)
+}
+
+func newStorager(pairs ...types.Pair) (store *Storage, err error) {
+	defer func() {
+		if err != nil {
+			err = services.InitError{Op: "new_storager", Type: Type, Err: formatError(err), Pairs: pairs}
+		}
+	}()
+
+	store = &Storage{
+		connection: nil,
+		user:       "anonymous",
+		password:   "anonymous",
+		url:        "localhost:21",
+		workDir:    "/",
+	}
+
+	opt, err := parsePairStorageNew(pairs)
+	if err != nil {
+		return
+	}
+
+	if opt.HasEndpoint {
+		ep, err := endpoint.Parse(opt.Endpoint)
+		if err != nil {
+			return nil, err
+		}
+		var host string
+		var port int
+		switch ep.Protocol() {
+		case endpoint.ProtocolHTTP:
+			_, host, port = ep.HTTP()
+		default:
+			return nil, services.PairUnsupportedError{Pair: ps.WithEndpoint(opt.Endpoint)}
+		}
+		url := fmt.Sprintf("%s:%d", host, port)
+		store.url = url
+	}
+
+	if opt.HasWorkDir {
+		store.workDir = opt.WorkDir
+	}
+
+	// Prepare for new protocol Basic
+	if opt.HasCredential {
```

Comment:
It's OK for me to remove this check temporarily, and added back in another PR when protocol Basic is supported.

Reply:
Or you can just leave it here, add `protocol Basic` support, and then we go back to uncomment this before this PR is merged.

Reply:
I'll add it in a divided PR.